vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java

```diff
@@ -572,13 +573,20 @@ protected boolean isInputValuePresent() {
     @Override
     public void setValue(LocalDate value) {
         LocalDate oldValue = getValue();
+        boolean isInputValuePresent = isInputValuePresent();
+
+        // When the value is cleared programmatically, reset hasInputValue
+        // so that the following validation doesn't treat this as bad input.
+        if (Objects.equals(value, getEmptyValue())) {
```

Comment:
**note:** Even though the issue doesn't manifest in DatePicker because its web component handles hasInputValue a bit differently, I've decided that it won't hurt to include the fix in DatePicker as well to prevent potential regressions if the implementation of the web component changes.